### ISSUE TITLE
Clip raster by mask layer gives different results in different versions

### ISSUE THREAD

**Issue by bananaonthemoon** [qgis/QGIS#34929]
**Describe the bug**

tl:dr it's possible that newer versions of "extract raster by mask" extract sliiiighly differently to previous versions (i checked 3.2, 3.10 and 3.12)?

During creating a lego map in 3.2, 3.10 and 3.12 (I had workshops, so everyone had different versions, I use 3.2 90% of the time) I noticed that 3.10 and 3.12 gives different results, even though I'd recreated the map at least 3 times using each version (and on different computers).  I tracked the first discrepancy back to the layer that was created after using "extract raster by mask". The extent of layer is slighly different, but after adding the grid and using the zonal statistics algorithm (mean) the max value between these versions are HUGE.
1) I thought that maybe the zonal statistics algorithm works differently in 3.10/3.12, but I tried the data i created with 3.2 and it worked fine (although it adds 0s instead of nulls, hmm)
2) I thought that maybe the coordinates were different, but everything was ok
3) as you will see in the pdf, for version 3.2 the clipped raster and the mask used have the same extents, but that's not the case for 3.12 (although I checked it, too)

**How to Reproduce**

I used 5x5 tiles that cover poland from srtm.csi.cgiar.org (there are 5 tiles just in case) (its over 600mb so i chose not to add it here)

the shp for administrative regions is here 
[voivodeships.zip](https://github.com/qgis/QGIS/files/4302433/voivodeships.zip)

I made a pdf of all the steps for 3.2 and 3.12
[report.pdf](https://github.com/qgis/QGIS/files/4302435/report.pdf)

**QGIS and OS versions**

![image](https://user-images.githubusercontent.com/61518849/76151385-f8a27e80-60b4-11ea-9fe7-63c09f7dd7d1.png)

![image](https://user-images.githubusercontent.com/61518849/76150727-ca6d7080-60ad-11ea-948b-e4b59dd6e7eb.png)

It took me like a week to make sure it's not just my fault or sloppiness, I hope this gets solved quickly :)

**Comment by nyalldawson** [comment]
This is a gdal based algorithm. It's likely the difference stems from a change in gdal itself. 

Check the command listed in the algorithm dialog for differences between the versions, but if they are identical then the change has come from gdal.

**Comment by gioman** [comment]
@bananaonthemoon as @nyalldawson says you must check this directly with GDAL:

in your PDF you'll see that in the images of the logs you see a "gdal command" and following the actual GDAL command that QGIS built and launched.


By the way:

Input raster layer metadata as returned by GDAL 3 on QGIS 3.10.3


```
Name	poland_int16_v2
Path	C:\Users\qgis\Desktop\poland_int16_v2.tif
CRS	EPSG:4326 - WGS 84 - Geographic
Extent	10.0000000000000000,45.0000000000000000 : 25.0000000000000000,55.0000000000000000
Unit	degrees
Width	18000
Height	12000
Data type	Int16 - Sixteen bit signed integer
GDAL Driver Description	GTiff
GDAL Driver Metadata	GeoTIFF
Dataset Description	C:/Users/qgis/Desktop/poland_int16_v2.tif
Compression	
Band 1	
STATISTICS_APPROXIMATE=YES
STATISTICS_MAXIMUM=2431
STATISTICS_MEAN=247.36301009916
STATISTICS_MINIMUM=-21
STATISTICS_STDDEV=248.09330507524
STATISTICS_VALID_PERCENT=78.86
More information	
AREA_OR_POINT=Area
Dimensions	X: 18000 Y: 12000 Bands: 1
Origin	10,55
Pixel Size	0.0008333333333333333868,-0.0008333333333333333868

```


Input raster layer metadata as returned by GDAL 2 on QGIS 3.2


```
Original | poland_int16_v2
-- | --
Name | poland_int16_v2
Source | C:/Users/qgis/Desktop/poland_int16_v2.tif
Provider | gdal
CRS | EPSG:4326 - WGS 84 - Geographic
Extent | 10.0000000000000000,45.0000000000000000 : 25.0000000000000000,55.0000000000000000
Unit | degrees
Width | 18000
Height | 12000
Data type | Int16 - Sixteen bit signed integer
GDAL Driver Description | GTiff
GDAL Driver Metadata | GeoTIFF
Dataset Description | C:/Users/qgis/Desktop/poland_int16_v2.tif
Compression |  
Band 1 | STATISTICS_MAXIMUM=2431 STATISTICS_MEAN=247.36301009916 STATISTICS_MINIMUM=-21 STATISTICS_STDDEV=248.09330507524
More information | AREA_OR_POINT=Area
Dimensions | X: 18000 Y: 12000 Bands: 1
Origin | 10,55
Pixel Size | 0.000833333,-0.000833333
```


They seems identical.


GDAL clip by mask command on QGIS 3.10:

`gdalwarp -ot Int16 -of GTiff -cutline C:\Users\qgis\Desktop\poland.gpkg -cl poland -crop_to_cutline C:/Users/qgis/Desktop/poland_int16_v2.tif C:/Users/qgis/Desktop/clipped_3_10_3.tif`


GDAL clip by mask command on QGIS 3.2:

`gdalwarp -ot Int16 -of GTiff -cutline C:/Users/qgis/Desktop/poland.gpkg -crop_to_cutline C:/Users/qgis/Desktop/poland_int16_v2.tif C:/Users/qgis/Desktop/clipped_3_2.tif`

The two command are identical.


GDAL metadata of result in QGIS 3.10.3:

```
Name	clipped_3_10_3
Path	C:\Users\qgis\Desktop\clipped_3_10_3.tif
CRS	EPSG:4326 - WGS 84 - Geographic
Extent	14.1233333333333348,49.0024999999999977 : 24.1450000000000031,54.8358333333333334
Unit	degrees
Width	12026
Height	7000
Data type	Int16 - Sixteen bit signed integer
GDAL Driver Description	GTiff
GDAL Driver Metadata	GeoTIFF
Dataset Description	C:/Users/qgis/Desktop/clipped_3_10_3.tif
Compression	
Band 1	
STATISTICS_APPROXIMATE=YES
STATISTICS_MAXIMUM=2154
STATISTICS_MEAN=169.42108641613
STATISTICS_MINIMUM=-13
STATISTICS_STDDEV=128.86317362441
STATISTICS_VALID_PERCENT=69.34
More information	
AREA_OR_POINT=Area
Dimensions	X: 12026 Y: 7000 Bands: 1
Origin	14.1233,54.8358
Pixel Size	0.0008333333333333334952,-0.0008333333333333337121
```

GDAL metadata of result in QGIS 3.2:


```
Original | Clipped (mask)
-- | --
Name | Clipped (mask)
Source | C:/Users/qgis/Desktop/clipped_3_2.tif
Provider | gdal
CRS | EPSG:4326 - WGS 84 - Geographic
Extent | 14.1228848600000561,49.0020465180000429 : 24.1457830750000539,54.8364166670000373
Unit | degrees
Width | 12027
Height | 7001
Data type | Int16 - Sixteen bit signed integer
GDAL Driver Description | GTiff
GDAL Driver Metadata | GeoTIFF
Dataset Description | C:/Users/qgis/Desktop/clipped_3_2.tif
Compression |  
Band 1 | STATISTICS_MAXIMUM=2154 STATISTICS_MEAN=169.42199941311 STATISTICS_MINIMUM=-7 STATISTICS_STDDEV=128.90011638951
More information | AREA_OR_POINT=Area
Dimensions | X: 12027 Y: 7001 Bands: 1
Origin | 14.1229,54.8364
Pixel Size | 0.000833366,-0.000833362
```



There is a difference for example the WxH size is slightly different, and it seems that is caused by the clipping operation because if I load in QGIS 3.10.3 the layer clipped in 3.2 the WxH is 12027x7001. I'm not sure if this is enough to justify the different results you are seeing.

Anyway here following the zonal stats results obtained in 3.10.3 and 3.2, results seems pretty similar to me:


https://fil.email/S7HDCH72

**Comment by bananaonthemoon** [comment]
Hi :)

the zonal stats is used on the grid layer, so it looks later like the lego map. in the 2nd pdf you can see the difference it makes

heres the pdf with my data +pdf
https://fil.email/kSHTNTcw
[zonal_stats.pdf](https://github.com/qgis/QGIS/files/4306649/zonal_stats.pdf)

after removing nulls (or in case of 3.10+ it should be 0s) and styling it should look like this 
![map-correct](https://user-images.githubusercontent.com/61518849/76215168-70d08780-620e-11ea-9b3d-61b0b104aa97.png)


and not like

![87359333_874227169698962_1024427258615431168_n](https://user-images.githubusercontent.com/61518849/76214688-8ee9b800-620d-11ea-9941-cb9d09f88050.png)

**Comment by gioman** [comment]
@bananaonthemoon is really hard to compare maps like the one you posted and added to the pdf, especially if you don't add a project with the symology used and the symbology classes. Let's try stick by the numbers, that won't fail :)

in the data you attached you have a vector grid that has a squares of 15000 meters of WxH, then you have a raster with 71.8m of resolution. For each vector square this means ~43681 pixels, that is what you get with the "count" operation in QGIS 3.10/3.12. I would say that the issue seems to be in older QGIS releases if you don't get the same values.

**Comment by bananaonthemoon** [comment]
The count number changes though? Plus in the north there are mountains (the carpathians) so there's no logic in low values there :(

here are projects for 3.2 and 3.12 https://fil.email/5CCeDDJw

**Comment by gioman** [comment]
> The count number changes though? Plus in the north there are mountains (the carpathians) so there's no logic in low values there :(

@bananaonthemoon

"count" is the number of pixels per polygon. If you have NULLs then the count within a polygon can change.

**Comment by bananaonthemoon** [comment]
I still think its 3.12. lets look at bordering polygons. most of them they barely overlays the raster but they the count number is still very high.

example:
![image](https://user-images.githubusercontent.com/61518849/76446282-1dee0000-63c7-11ea-9a0d-3d33051383a9.png)
3.2
![image](https://user-images.githubusercontent.com/61518849/76446318-2e9e7600-63c7-11ea-9a86-6390e4150b2d.png)
3.12
![image](https://user-images.githubusercontent.com/61518849/76446354-3d852880-63c7-11ea-913e-a6269f3245c7.png)

**Comment by gioman** [comment]
@bananaonthemoon as you can see

![Screenshot_20200311_192049](https://user-images.githubusercontent.com/1951107/76455353-90fa7500-63cd-11ea-8f02-fb3bcb9a6bc0.png)

on the edge of your grid, your merged raster as 0 values, not NULLs. Those are valid pixels, so the are part of the "count" stat. You can also see that the original rasters have NULLs (no data) in the same spot. This means that when merging the raster you must keep the no data outside their boundaries. Then you will get the "count" and "sum", etc. stats you expect.

I think this can be closed.

**Comment by alexbruy** [comment]
@bananaonthemoon can we close this?

**Comment by bananaonthemoon** [comment]
Last question, how do I merge the raster and keep the no data outside their boundaries at the same time?

**Comment by gioman** [comment]
> Last question, how do I merge the raster and keep the no data outside their boundaries at the same time?

@bananaonthemoon In your case the necessary caution must be taken with you clip the merged raster with the boundaries:

![Screenshot_20200313_173119](https://user-images.githubusercontent.com/1951107/76646165-6a197b80-6552-11ea-91dc-16c8b10aa628.png)

I think you must set it to 0, it means that pixels that would otherwise be outputted as 9 will be classified as NULLs.
